Support In and notIn operators in ParquetFilters.ConvertFilterToParquet

sririshindra opened the pull request:
ParquetFilters.ConvertFilterToParquet performs conversion from an Iceberg filter (Expression) to a Parquet FilterPredicate. It currently only handles IS_NULL, NOT_NULL, IS_NAN, NOT_NAN, EQ, NOT_EQ, GT, GT_EQ, LT, LT_EQ. It does not handle, e.g., IN and NOT_IN, which would be useful to have.

This conversion can be used for setting a record filter in Parquet and then using that to filter row groups when reading Parquet files.

sririshindra commented:
I originally created https://github.com/apache/iceberg/pull/10341 . It did not get traction at the time and it was closed. I would appreciate if the committers can take another look at this as we think its a legitimate improvement. cc: @wypoon

wypoon commented on parquet/src/test/java/org/apache/iceberg/parquet/TestParquetFilters.java:
```diff
@@ -0,0 +1,86 @@
+@ExtendWith(ParameterizedTestExtension.class)
+public class TestParquetFilters {
+
+  private static final java.util.Map<String, Integer> ALIASES =
+      new java.util.HashMap<String, Integer>() {
+        {
+          put("id", 1);
+          put("age", 2);
+        }
+      };
```
In Iceberg, the preference is to avoid instantiating something like `java.util.HashMap` or `java.util.ArrayList` directly. You can use `com.google.common.collect.ImmutableMap` here instead. `ImmutableMap.of("id", 1, "age", 2)` would work.

sririshindra replied:
Fixed in latest commit.

wypoon commented on parquet/src/test/java/org/apache/iceberg/parquet/TestParquetFilters.java:
```diff
@@ -0,0 +1,86 @@
+        (FilterCompat.Filter) ParquetFilters.convert(SCHEMA, in("id", 1, 2, 3), true);
+
+    try {
+      java.lang.reflect.Field privateField = filter.getClass().getDeclaredField("filterPredicate");
+      privateField.setAccessible(true);
+      assertThat(privateField.get(filter).toString().equalsIgnoreCase("in(id, 1, 2, 3)")).isTrue();
```
Since you do this more than once, you can consider a helper method to get the private `Field`, make it accessible, and then get the filter from it and doing a `toString()` on that.

sririshindra replied:
Extracted it to a seperate method.

wypoon reviewed (commented):
@sririshindra thanks for this PR.
Do you think it is worth adding tests where the column is a long/timestamp and where the column is a string (binary column)?

wypoon commented on api/src/main/java/org/apache/iceberg/Schema.java:
```diff
@@ -204,6 +204,10 @@ public Schema(int schemaId, NestedField... columns) {
     this(schemaId, Arrays.asList(columns));
   }
 
+  public Schema(Map<String, Integer> aliases, NestedField... columns) {
+    this(Arrays.asList(columns), aliases);
+  }
```
I'd be hesitant to add to the API. This is only used for a test; can't you use an existing constructor?

sririshindra replied:
I removed it in the latest version.

sririshindra commented:
> @sririshindra thanks for this PR. Do you think it is worth adding tests where the column is a long/timestamp and where the column is a string (binary column)?

Thanks @wypoon I added a test for long and string as well. That actually turned out to be very useful as that caught a bug that I didn't know previously. There was an unnecessary cast of the litset to HashSet in the previous version. I now removed it and the Tests that I added all succeed now.

wypoon commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetFilters.java:
```diff
@@ -118,11 +119,14 @@ public <T> FilterPredicate predicate(BoundPredicate<T> pred) {
         lit = null;
       } else if (pred.isLiteralPredicate()) {
         lit = pred.asLiteralPredicate().literal();
+      } else if (pred.isSetPredicate()) {
+        litSet = pred.asSetPredicate().asSetPredicate().literalSet();
```
```
        litSet = pred.asSetPredicate().literalSet();
```
(You call `asSetPredicate()` twice.)

sririshindra replied:
Fixed in latest commit.

wypoon reviewed (commented):
LGTM otherwise.

wypoon commented:
@rdblue can you review this?
I have some stale PRs implementing Parquet page skipping and at that time I came across the fact that `ParquetFilters.ConvertFilterToParquet` does not convert `IN`, for example.

adrians commented:
Bump! This feature would simplify the query-generation logic in a client-app.
